File: coordinator/safety.py
# -*- coding: utf-8 -*-
"""
安全阈值模块 (Safety Thresholds)
================================
分布式算力系统的"刹车"：任何设备(PC/手机)在贡献算力前，都要先过安全阈值检查。
超阈值 → 拒绝新任务 / 暂停 / 降速，保护设备不被算力租用拖垮、不过热、不耗尽电量。

阈值分三档：
  GREEN  正常       —— 可接全速任务
  YELLOW 接近上限   —— 只接小任务 / 降速
  RED    超阈值     —— 拒绝一切新任务，已有任务尽快收尾

PC 端用 psutil 读真实指标；手机端的阈值在 Android worker 里同构实现。
"""
import os

try:
    import psutil
except ImportError:
    psutil = None


# ── 默认阈值（可被环境变量 / data 覆盖）──────────────────────────────────────
DEFAULTS = {
    # CPU 占用率上限（%）。超过 RED 不再派活。
    "cpu_yellow": 75.0,
    "cpu_red":    90.0,
    # 内存占用率上限（%）
    "mem_yellow": 80.0,
    "mem_red":    92.0,
    # 磁盘占用率上限（%，针对租出去的盘）
    "disk_yellow": 85.0,
    "disk_red":    95.0,
    # 温度上限（℃，PC 多数无传感器时跳过；手机必查）
    "temp_yellow": 70.0,
    "temp_red":    82.0,
    # 电池下限（%，仅手机/笔记本）。低于 RED 停止贡献。
    "battery_red": 20.0,
    "battery_yellow": 35.0,
    # 是否要求充电中才贡献（手机强烈建议 True）
    "require_charging": False,
    # 网络流量上限（本次会话 MB；超了停网络类任务，防跑爆流量套餐）
    "net_session_mb_red": 5000.0,
    # 预留给系统的最小空闲内存（MB），低于此一律 RED。
    # 注：手机/小内存设备(如1GB模拟器)按绝对值1GB太苛刻，降到256MB；
    #   真正的内存保护交给 mem_red(占用率%) 这个相对指标。
    "min_free_mem_mb": 256.0,
}


def load_thresholds(overrides=None):
    t = dict(DEFAULTS)
    if overrides:
        t.update({k: v for k, v in overrides.items() if k in t})
    # 环境变量覆盖：SAFE_CPU_RED=85 等
    for k in t:
        env = os.environ.get("SAFE_" + k.upper())
        if env is not None:
            try:
                t[k] = type(t[k])(env) if not isinstance(t[k], bool) else env == "1"
            except Exception:
                pass
    return t


# ── 读取本机真实资源 ────────────────────────────────────────────────────────
def read_resources():
    """返回当前 PC 资源快照。手机端在 Android 侧用 BatteryManager/ActivityManager 同构。"""
    r = {"cpu": 0.0, "mem": 0.0, "mem_free_mb": 99999.0, "disk": 0.0,
         "temp": None, "battery": None, "charging": None,
         "cores": 1, "mem_total_mb": 0.0}
    if not psutil:
        return r
    try:
        r["cpu"] = psutil.cpu_percent(interval=0.3)
        r["cores"] = psutil.cpu_count() or 1
        vm = psutil.virtual_memory()
        r["mem"] = vm.percent
        r["mem_free_mb"] = round(vm.available / 1048576, 1)
        r["mem_total_mb"] = round(vm.total / 1048576, 1)
        # 系统盘占用
        du = psutil.disk_usage(os.path.abspath(os.sep))
        r["disk"] = du.percent
        # 温度（很多 PC 无此传感器）
        try:
            temps = psutil.sensors_temperatures()
            allt = [s.current for arr in temps.values() for s in arr if s.current]
            if allt:
                r["temp"] = round(max(allt), 1)
        except Exception:
            pass
        # 电池（笔记本有，台式无）
        try:
            bat = psutil.sensors_battery()
            if bat is not None:
                r["battery"] = round(bat.percent, 1)
                r["charging"] = bool(bat.power_plugged)
        except Exception:
            pass
    except Exception:
        pass
    return r


# ── 阈值判定 ────────────────────────────────────────────────────────────────
def evaluate(resources, thresholds, net_session_mb=0.0):
    """
    返回 (level, reasons)
      level ∈ {"GREEN","YELLOW","RED"}
      reasons = 触发原因列表
    """
    t = thresholds
    reasons = []
    level = "GREEN"

    def bump(new_level, why):
        nonlocal level
        order = {"GREEN": 0, "YELLOW": 1, "RED": 2}
        if order[new_level] > order[level]:
            level = new_level
        reasons.append(why)

    cpu = resources.get("cpu", 0)
    if cpu >= t["cpu_red"]:   bump("RED",    "CPU %.0f%% ≥ %.0f%%" % (cpu, t["cpu_red"]))
    elif cpu >= t["cpu_yellow"]: bump("YELLOW", "CPU %.0f%%" % cpu)

    mem = resources.get("mem", 0)
    if mem >= t["mem_red"]:   bump("RED",    "内存 %.0f%% ≥ %.0f%%" % (mem, t["mem_red"]))
    elif mem >= t["mem_yellow"]: bump("YELLOW", "内存 %.0f%%" % mem)

    free = resources.get("mem_free_mb", 99999)
    if free < t["min_free_mem_mb"]:
        bump("RED", "空闲内存仅 %.0fMB < %.0fMB" % (free, t["min_free_mem_mb"]))

    disk = resources.get("disk", 0)
    if disk >= t["disk_red"]:   bump("RED",    "磁盘 %.0f%% ≥ %.0f%%" % (disk, t["disk_red"]))
    elif disk >= t["disk_yellow"]: bump("YELLOW", "磁盘 %.0f%%" % disk)

    temp = resources.get("temp")
    if temp is not None:
        if temp >= t["temp_red"]:   bump("RED",    "温度 %.0f℃ ≥ %.0f℃" % (temp, t["temp_red"]))
        elif temp >= t["temp_yellow"]: bump("YELLOW", "温度 %.0f℃" % temp)

    bat = resources.get("battery")
    charging = resources.get("charging")
    if bat is not None:
        if bat <= t["battery_red"]:   bump("RED",    "电量 %.0f%% ≤ %.0f%%" % (bat, t["battery_red"]))
        elif bat <= t["battery_yellow"]: bump("YELLOW", "电量 %.0f%%" % bat)
    if t.get("require_charging") and charging is False:
        bump("RED", "未充电(策略要求充电中才贡献)")

    if net_session_mb >= t["net_session_mb_red"]:
        bump("RED", "本会话流量 %.0fMB ≥ %.0fMB" % (net_session_mb, t["net_session_mb_red"]))

    if not reasons:
        reasons.append("全部指标正常")
    return level, reasons


def can_accept_job(resources, thresholds, net_session_mb=0.0, job_weight="normal"):
    """
    是否可接新任务。
      job_weight: "small" | "normal" | "heavy"
      GREEN  接全部
      YELLOW 只接 small
      RED    一律拒绝
    """
    level, reasons = evaluate(resources, thresholds, net_session_mb)
    if level == "RED":
        return False, level, reasons
    if level == "YELLOW" and job_weight != "small":
        return False, level, reasons + ["YELLOW 档只接 small 任务"]
    return True, level, reasons


if __name__ == "__main__":
    th = load_thresholds()
    res = read_resources()
    lvl, why = evaluate(res, th)
    print("资源:", res)
    print("阈值档位:", lvl)
    print("原因:", "; ".join(why))

File: worker-android/app/src/main/java/com/example/cpu/Safety.kt
package com.example.cpu

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import java.io.RandomAccessFile

/**
 * 手机端安全阈值 + 资源读取（与 PC safety.py 同构）。
 *
 * 手机贡献算力前必须过安全阈值：电量低 / 未充电 / 高温 / 内存紧张 → 拒绝任务。
 * 这是保护你手机的"刹车"——算力租用绝不能把手机跑到过热、耗光电、卡死。
 */
data class Resources(
    var cpu: Double = 0.0,        // CPU 占用率 %
    var mem: Double = 0.0,        // 内存占用率 %
    var memFreeMb: Double = 0.0,
    var memTotalMb: Double = 0.0,
    var cores: Int = 1,
    var battery: Int = 100,       // 电量 %
    var charging: Boolean = false,
    var temp: Double? = null,     // 电池温度 ℃
)

data class Thresholds(
    val cpuYellow: Double = 75.0, val cpuRed: Double = 90.0,
    val memYellow: Double = 80.0, val memRed: Double = 92.0,
    val tempYellow: Double = 38.0, val tempRed: Double = 44.0,  // 手机电池温度，比PC低
    val batteryRed: Int = 20, val batteryYellow: Int = 35,
    val requireCharging: Boolean = false,
    val minFreeMemMb: Double = 256.0,
)

enum class Level { GREEN, YELLOW, RED }

object Safety {

    /** 读取手机实时资源。 */
    fun read(ctx: Context): Resources {
        val r = Resources()
        r.cores = Runtime.getRuntime().availableProcessors()

        // 内存：ActivityManager.MemoryInfo
        val am = ctx.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val mi = ActivityManager.MemoryInfo()
        am.getMemoryInfo(mi)
        r.memTotalMb = mi.totalMem / 1048576.0
        r.memFreeMb = mi.availMem / 1048576.0
        if (r.memTotalMb > 0) r.mem = (1.0 - mi.availMem.toDouble() / mi.totalMem) * 100.0

        // 电池：sticky broadcast
        try {
            val bm = ctx.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
            r.battery = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
            val status = ctx.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            if (status != null) {
                val st = status.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
                r.charging = st == BatteryManager.BATTERY_STATUS_CHARGING ||
                        st == BatteryManager.BATTERY_STATUS_FULL
                val tempRaw = status.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1)
                if (tempRaw > 0) r.temp = tempRaw / 10.0  // 单位 0.1℃
            }
        } catch (_: Exception) {}

        // CPU 占用：读 /proc/stat 两次差分（部分新系统受限，失败则 0）
        r.cpu = readCpuPercent()
        return r
    }

    private var lastIdle = 0L
    private var lastTotal = 0L
    private fun readCpuPercent(): Double {
        return try {
            RandomAccessFile("/proc/stat", "r").use { raf ->
                val line = raf.readLine() ?: return 0.0
                val toks = line.split(Regex("\\s+")).drop(1).mapNotNull { it.toLongOrNull() }
                if (toks.size < 4) return 0.0
                val idle = toks[3]
                val total = toks.sum()
                val dIdle = idle - lastIdle
                val dTotal = total - lastTotal
                lastIdle = idle; lastTotal = total
                if (dTotal <= 0) 0.0 else ((1.0 - dIdle.toDouble() / dTotal) * 100.0).coerceIn(0.0, 100.0)
            }
        } catch (_: Exception) { 0.0 }
    }

    /** 阈值判定 → (档位, 原因列表)。 */
    fun evaluate(r: Resources, t: Thresholds): Pair<Level, List<String>> {
        var level = Level.GREEN
        val reasons = mutableListOf<String>()
        fun bump(l: Level, why: String) {
            if (l.ordinal > level.ordinal) level = l
            reasons.add(why)
        }
        if (r.cpu >= t.cpuRed) bump(Level.RED, "CPU ${r.cpu.toInt()}% ≥ ${t.cpuRed.toInt()}%")
        else if (r.cpu >= t.cpuYellow) bump(Level.YELLOW, "CPU ${r.cpu.toInt()}%")

        if (r.mem >= t.memRed) bump(Level.RED, "内存 ${r.mem.toInt()}% ≥ ${t.memRed.toInt()}%")
        else if (r.mem >= t.memYellow) bump(Level.YELLOW, "内存 ${r.mem.toInt()}%")

        if (r.memFreeMb < t.minFreeMemMb) bump(Level.RED, "空闲内存仅 ${r.memFreeMb.toInt()}MB")

        r.temp?.let {
            if (it >= t.tempRed) bump(Level.RED, "电池温度 ${it}℃ ≥ ${t.tempRed}℃(过热保护)")
            else if (it >= t.tempYellow) bump(Level.YELLOW, "电池温度 ${it}℃")
        }
        if (r.battery <= t.batteryRed) bump(Level.RED, "电量 ${r.battery}% ≤ ${t.batteryRed}%")
        else if (r.battery <= t.batteryYellow) bump(Level.YELLOW, "电量 ${r.battery}%")

        if (t.requireCharging && !r.charging) bump(Level.RED, "未充电(策略要求充电才贡献)")

        if (reasons.isEmpty()) reasons.add("全部正常")
        return level to reasons
    }

    /** 是否可接任务。RED 一律拒；YELLOW 只接 small。 */
    fun canAccept(r: Resources, t: Thresholds, weight: String): Boolean {
        val (lvl, _) = evaluate(r, t)
        return when (lvl) {
            Level.RED -> false
            Level.YELLOW -> weight == "small"
            Level.GREEN -> true
        }
    }
}
